sdk/dapp-kit/src/components/SuiClientProvider.tsx

```diff
@@ -2,29 +2,72 @@
 // SPDX-License-Identifier: Apache-2.0
 
 import { SuiClient, getFullnodeUrl } from '@mysten/sui.js/client';
-import { SuiClientContext } from '../hooks/useSuiClient.js';
-import { useMemo } from 'react';
+import type { SuiClientOptions } from '@mysten/sui.js/client';
+import { createContext, useMemo, useState } from 'react';
 
-export interface SuiClientProviderProps {
+type NetworkConfig = SuiClient | SuiClientOptions;
+type NetworkConfigs<T extends NetworkConfig = NetworkConfig> = Record<string, T>;
+
+export interface SuiClientProviderContext {
+	client: SuiClient;
+	queryKey: (key: unknown[]) => unknown[];
+	networks: NetworkConfigs;
+	selectNetwork: (network: string) => void;
+}
+
+export const SuiClientContext = createContext<SuiClientProviderContext | undefined>(undefined);
+
+export interface SuiClientProviderProps<T extends NetworkConfigs> {
+	networks?: T;
+	createClient?: (name: keyof T, config: T[keyof T]) => SuiClient;
+	defaultNetwork?: keyof T;
 	children: React.ReactNode;
-	client?: SuiClient;
-	url?: string;
-	queryKeyPrefix: string;
 }
 
-export const SuiClientProvider = (props: SuiClientProviderProps) => {
-	const ctx = useMemo(() => {
-		const client =
-			props.client ??
-			new SuiClient({
-				url: props.url ?? getFullnodeUrl('devnet'),
-			});
+export function SuiClientProvider<T extends NetworkConfigs>(props: SuiClientProviderProps<T>) {
+	const networks = useMemo(
+		() =>
+			props.networks ??
+			({
+				devnet: { url: getFullnodeUrl('devnet') },
+			} as unknown as T),
+		[props.networks],
+	);
+	const [selectedNetwork, setSelectedNetwork] = useState<keyof T>(
+		props.defaultNetwork ?? Object.keys(networks)[0],
+	);
+
+	const createClient = useMemo(() => {
+		if (props.createClient) {
+			return props.createClient;
+		}
 
+		return (_name: keyof T, config: T[keyof T]) => {
+			if (config instanceof SuiClient) {
+				return config;
+			}
+
+			return new SuiClient(config);
+		};
+	}, [props.createClient]);
+
+	const [client, setClient] = useState<SuiClient>(() => {
+		return createClient(selectedNetwork, networks[selectedNetwork]);
+	});
+
+	const ctx = useMemo((): SuiClientProviderContext => {
 		return {
 			client,
-			queryKey: (key: unknown[]) => [props.queryKeyPrefix, ...key],
+			queryKey: (key: unknown[]) => [selectedNetwork, ...key],
+			networks,
+			selectNetwork: (network: keyof T) => {
+				if (network !== selectedNetwork) {
+					setSelectedNetwork(network);
+					setClient(createClient(network, networks[network]));
```

Comment:
something I am unsure about it if the clients should be cached as they are created.  We don't currently do any cleanup.  Currently the only thing persisted in the clients are websocket connections for active subscriptions.  I could imagine just calling a close method whenever we switch...

Reply:
Yeah that's a great question. I actually don't think that we need to do any caching because you'd just be able to do cleanup in an effect. e.g.:

```ts
useEffect(() => {
  return () => client.cleanup();
}, [client]);

Reply:
would it make more sense to clean up the previous client imperatively in `selectNetwork` versus cleaning up the client as a side effect of rendering?

Reply:
Going to leave cleanup question for the future when there is actually something to clean up, since clients currently don't support this